identity/src/commonMain/kotlin/com/android/identity/nfc/NdefRecord.kt

```diff
@@ -0,0 +1,355 @@
+
+        /**
+         * Indicates the type field contains an external type name.
+         */
+        EXTERNAL_TYPE(0x04),
+
+        /**
+         * Indicates the type field is unknown.
+         */
+        UNKNOWN(0x05),
+
+        /**
+         * Indicates the payload is an intermediate or final chunk of a chunked NDEF Record.
+         */
+        UNCHANGED(0x06),
+
+        /**
+         * Reserved TNF type.
+         */
+        RESERVED(0x07)
+    }
+
+    /**
+     * Encodes the record into a [Buffer].
+     *
+     * @param buf the [Buffer] to append the encoding of the record to.
+     * @param mb set to `true` if this is the first record of a message.
+     * @param me set to `true` if this is the last record of a message.
+     */
+    internal fun encode(
+        buf: Buffer,
+        mb: Boolean,
+        me: Boolean
+    ) {
+        val sr = payload.size < 256
+        val il = if (tnf == Tnf.EMPTY) true else (id.size > 0)
+        var flags = tnf.value
+        if (mb) flags = flags or FLAG_MB
+        if (me) flags = flags or FLAG_ME
+        if (sr) flags = flags or FLAG_SR
+        if (il) flags = flags or FLAG_IL
+
+        buf.writeByte(flags.toByte())
+        buf.writeByte(type.size.toByte())
+        if (sr) {
+            buf.writeByte(payload.size.toByte())
+        } else {
+            buf.writeInt(payload.size)
+        }
+        if (il) {
+            buf.writeByte(id.size.toByte())
+        }
+        buf.write(type.toByteArray())
+        buf.write(id.toByteArray())
+        buf.write(payload.toByteArray())
+    }
+
+    /**
+     * The URI of record, if one can be found.
+     *
+     * The following records are supported:
+     * - Records with [Tnf.ABSOLUTE_URI]
+     * - [Tnf.WELL_KNOWN] with type [Nfc.RTD_URI]
+     * - [Tnf.WELL_KNOWN] with type [Nfc.RTD_SMART_POSTER] and containing a URI record in the NDEF message nested in the
+     * payload.
+     *
+     * If this is not a URI record by the above rules, `null` is returned.
+     */
+    val uri: String?
+        get() = getUri(false)
+
+    private fun getUri(inSmartPoster: Boolean): String? {
+        return when (tnf) {
+            Tnf.WELL_KNOWN -> {
+                when (type) {
+                    Nfc.RTD_SMART_POSTER -> {
+                        try {
+                            val nestedMessage = NdefMessage.fromEncoded(payload.toByteArray())
+                            for (record in nestedMessage.records) {
+                                val uri = record.getUri(true)
+                                if (uri != null) {
+                                    return uri
+                                }
+                            }
+                            return null
+                        } catch (e: Throwable) {
+                            return null
+                        }
+                    }
+                    Nfc.RTD_URI -> return URI_PREFIX_MAP[payload[0].toInt()] + payload.toByteArray(1).decodeToString()
```

Comment:
The map will throw an exception if we're passed an unrecognized or invalid value.

...Which raises more of an API question: Would we rather have the API throw exceptions for invalid NFC input, or log and return null? Curious to get your thoughts. IMHO an exception is fine as long as we handle it at a reasonable level (ie. either skipping the invalid packet or letting the user know the exchange failed, rather than crashing the whole app).

Reply:
I think it's absolutely reasonable to throw an exception here, this only happens when the `uri` property is read.